packages/stokhos/src/sacado/kokkos/vector/belos/Belos_PseudoBlockGmresIter_MP_Vector.hpp

```diff
@@ -0,0 +1,884 @@
+//@HEADER
+// ************************************************************************
+//
+//                 Belos: Block Linear Solvers Package
+//                  Copyright 2004 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+
+#ifndef BELOS_PSEUDO_BLOCK_GMRES_ITER_MP_VECTOR_HPP
+#define BELOS_PSEUDO_BLOCK_GMRES_ITER_MP_VECTOR_HPP
+
+/*! 
+  \file Belos_PseudoBlockGmresIter_MP_Vector.hpp
+  \brief Belos concrete class for performing the pseudo-block GMRES iteration.
+
+  Specialized for Sacado::MP::Vector scalar type to handle break-down when
+  ensemble components are over-solved.
+*/
+
+#include "Stokhos_Sacado_Kokkos_MP_Vector.hpp"
+#include "BelosPseudoBlockGmresIter.hpp"
+
+/*!
+  \class Belos::PseudoBlockGmresIter
```

Comment:
This is a more difficult question to answer.
At the bare minimum, we need some modifications of the member function updateLSQR.
However, as you know, we cannot do a template specialization of a member function alone. We need, at least, a partial template specialization of the class.

I see two approaches to reduce the number of duplicated lines here but both of them will impact Belos and, I think, should be included in another PR.

The first approach would be to remove updateLSQR from the PseudoBlockGmresIter class and only do a template specialization of the function.
This approach would work but is not very portable to other Belos solvers as it would, first, require to identify all the codes that can be reused or not.

The second approach would be to include the definition of the Mask and the MaskAssign classes (with templated Scalar type) in a package such as Teuchos. Then, the implementation of this PR (with the modification that you have suggested taken into account) could be used directly in Belos and there would just be a partial template specialization of the Mask and MaskAssign classes in Stokhos for ensemble types (as it is already the case).

I think that the second approach would be the best and should ease the maintenance of some specialized classes and functions in Stokhos that could just be not specialized anymore using Mask and MaskAssign as a type trait in some sense.

What are you thought about that?